lib/entity/event_list.rb

```diff
@@ -1,32 +1,62 @@
 module Synthea
-  class EventList < Array
+  class EventList
 
-    def select(&block)
-      EventList.new(super.select(&block))
+    EMPTY = [].freeze
+
+    attr_accessor :events     # Hash of Synthea::Event
+
+    def initialize
+      @events = {}
+      @events[:all] = []
+      @events[:unprocessed] = []
     end
-    
-    def next
-      self.first
+
+    def unprocessed
+      @events[:unprocessed]
     end
 
-    def next?
-      !self.empty?
+    def process(event)
+      event.processed = true
+      @events[:unprocessed].delete(event)
     end
 
-    def unprocessed
-      self.select{|x|x.processed==false}
+    def unprocessed_before(date,type)
+      @events[:unprocessed].select{|e|e.type==type && e.time <= date}
+    end
+
+    def unprocessed_since(date,type)
+      @events[:unprocessed].select{|e|e.type==type && e.time >= date}
     end
 
-    def before(date)
-      self.select{|x|x.time <= date}
+    def before(date,type=:all)
+      list = @events[type]
+      return EMPTY if list.nil? || list.empty?
+      end_index = list.index{|e|e.time >= date}
+      end_index = list.length if end_index.nil?
+      if !end_index.nil?
+        list[0..(end_index-1)]
```

Comment:
If a date is used that comes before any event in the list, then end_index is 0. Line 37: list[0..-1] may cause problems.

Also the else statement is never called because end_index will never be nil since it's set in the previous line.